aten/src/ATen/native/kleidiai/kai_kernels.cpp

```diff
@@ -0,0 +1,442 @@
+      if (m % num_thread_y == 0 && n % num_thread_x == 0) {
+        int64_t m_per_thread = m / num_thread_y;
+        int64_t n_per_thread = n / num_thread_x;
+        if (m_per_thread % gemm_m_step == 0 &&
+            n_per_thread % gemm_n_step == 0) {
+          division = {num_thread_x, num_thread_y, true};
+          return division;
+        }
+      }
+      max_threads -= 2;
+    }
+  }
+  // Split threads 1D for GEMV
+  if (n % gemv_n_step == 0) {
+    for (; max_threads > 0; max_threads -= 2) {
+      if (n % max_threads == 0 && (n / max_threads) % gemv_n_step == 0) {
+        division.num_threads_x = max_threads;
+        return division;
+      }
+    }
+  }
+  return division;
+}
+
+static void kai_quant_pack_lhs_int4_mm_groupwise(
+    const Tensor& output,
+    const Tensor& input,
+    const Tensor& weight,
+    const int64_t m,
+    const int64_t n,
+    const int64_t k,
+    const int64_t bl) {
+  kai_kernel_id id = kai_kernel_id::
+      matmul_clamp_f32_qai8dxp1x8_qsi4c32p8x8_1x8x32_neon_dotprod;
+  if (cpuinfo_has_arm_i8mm() && m > 1) {
+    id =
+        kai_kernel_id::matmul_clamp_f32_qai8dxp4x8_qsi4c32p4x8_4x8x32_neon_i8mm;
+  }
+  auto kernel_packet = kai_select_groupwise_matmul_ukernel(id);
+
+  const auto& ukernel = kernel_packet.ukernel;
+
+  const size_t mr = ukernel.get_mr();
+  const size_t kr = ukernel.get_kr();
+  const size_t sr = ukernel.get_sr();
+  const size_t n_step = ukernel.get_n_step();
+  int64_t total_threads = at::get_num_threads();
+  int64_t num_threads_x = 1;
+  adjust_max_threads(total_threads);
+  // Split threads 1D only for now
+  if (n % n_step == 0) {
+    for (; total_threads > 0; total_threads -= 2) {
+      if (n % total_threads == 0 && (n / total_threads) % n_step == 0) {
+        num_threads_x = total_threads;
+        break;
+      }
+    }
+  }
+
+  const size_t num_n_per_thread = n / num_threads_x;
+
+  const size_t dst_stride = n * sizeof(float);
+  float* lhs = reinterpret_cast<float*>(input.data_ptr());
+  uint8_t* rhs_packed_mtx_qs4cx = reinterpret_cast<uint8_t*>(weight.data_ptr());
+
+  uint8_t* dst_act_mtx_f32 = reinterpret_cast<uint8_t*>(output.data_ptr());
+  const size_t lhs_packed_size =
+      kernel_packet.kai_get_lhs_packed_size(m, k, mr, kr, sr);
+  uint8_t* lhs_packed = new uint8_t[lhs_packed_size];
+
+  // Quantize and pack the Input
+  kernel_packet.kai_run_lhs_quant_pack(
+      m,
+      k,
+      mr,
+      kr,
+      sr,
+      0,
+      (const float*)lhs,
+      k * sizeof(float),
+      (void*)lhs_packed);
+
+  at::parallel_for(0, num_threads_x, 0, [&](int begin, int end) {
+    for (int64_t x = begin; x < end; x++) {
+      matmul_groupwise(
+          std::ref(ukernel),
+          m,
+          num_n_per_thread,
+          x * num_n_per_thread,
+          k,
+          bl,
+          dst_stride,
+          lhs_packed,
+          rhs_packed_mtx_qs4cx,
+          dst_act_mtx_f32);
+    }
+  });
+
+  delete[] lhs_packed;
```

Comment:
It's 2024, all compiler support smart pointers, please use them
```suggestion
```

Reply:
done